relay-server/src/actors/healthcheck.rs

```diff
@@ -41,27 +58,39 @@ impl fmt::Display for SendError {
 
 impl Error for SendError {}
 
-/// Channel for sending public messages into a service.
+/// The address for a [`Service`].
 ///
-/// To send a message, use [`Addr::send`].
-#[derive(Clone, Debug)]
-pub struct Addr<T> {
-    tx: mpsc::UnboundedSender<Message<T>>,
+/// The address of a [`Service`] allows you to [send](Addr::send) messages to the service as
+/// long as the service is running.  It can be freely cloned.
+#[derive(Debug)]
+pub struct Addr<S: Service> {
+    tx: mpsc::UnboundedSender<S::Envelope>,
 }
 
-impl<T> Addr<T> {
+// For some reason the derive macro can't cope with this ¯\_(ツ)_/¯
+impl<S: Service> Clone for Addr<S> {
+    fn clone(&self) -> Self {
+        Self {
+            tx: self.tx.clone(),
+        }
+    }
+}
+
+impl<S: Service> Addr<S> {
     /// Sends an asynchronous message to the service and waits for the response.
     ///
     /// The result of the message does not have to be awaited. The message will be delivered and
     /// handled regardless. The communication channel with the service is unbounded, so backlogs
     /// could occur when sending too many messages.
     ///
     /// Sending the message can fail with `Err(SendError)` if the service has shut down.
-    pub async fn send(&self, data: T) -> Result<bool, SendError> {
-        let (responder, rx) = oneshot::channel();
-        let message = Message { data, responder };
-        self.tx.send(message).map_err(|_| SendError)?;
-        rx.await.map_err(|_| SendError)
+    pub async fn send<M>(&self, message: M) -> Result<M::Response, SendError>
```

Comment:
The returned future does not hold a reference to `self`. This is not actually required, the instance of the `Addr` can cease to exist while the future is pending for the response. From an API design perspective, this is too restrictive and can cause redundant clones of the address.

The [async book recommends](https://rust-lang.github.io/async-book/03_async_await/01_chapter.html#async-lifetimes) to return an `impl Future` in such cases.